datafusion/core/src/physical_optimizer/limit_pushdown.rs

```diff
@@ -0,0 +1,566 @@
+            "GlobalLimitExec: skip=0, fetch=5",
+            "  CoalescePartitionsExec",
+            "    LocalLimitExec: fetch=5",
+            "      FilterExec: c3@2 > 0",
+            "        RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=1",
+            "          StreamingTableExec: partition_sizes=1, projection=[c1, c2, c3], infinite_source=true"
+        ];
+        assert_eq!(get_plan_string(&after_optimize), expected);
+
+        Ok(())
+    }
+
+    fn create_schema() -> SchemaRef {
+        Arc::new(Schema::new(vec![
+            Field::new("c1", DataType::Int32, true),
+            Field::new("c2", DataType::Int32, true),
+            Field::new("c3", DataType::Int32, true),
+        ]))
+    }
+
+    fn streaming_table_exec(schema: SchemaRef) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(StreamingTableExec::try_new(
+            schema.clone(),
+            vec![Arc::new(DummyStreamPartition {
+                schema: schema.clone(),
+            }) as _],
+            None,
+            None,
+            true,
+            None,
+        )?))
+    }
+
+    fn global_limit_exec(
+        input: Arc<dyn ExecutionPlan>,
+        skip: usize,
+        fetch: Option<usize>,
+    ) -> Arc<dyn ExecutionPlan> {
+        Arc::new(GlobalLimitExec::new(input, skip, fetch))
+    }
+
+    fn local_limit_exec(
+        input: Arc<dyn ExecutionPlan>,
+        fetch: usize,
+    ) -> Arc<dyn ExecutionPlan> {
+        Arc::new(LocalLimitExec::new(input, fetch))
+    }
+
+    fn projection_exec(
+        schema: SchemaRef,
+        input: Arc<dyn ExecutionPlan>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(ProjectionExec::try_new(
+            vec![
+                (col("c1", schema.as_ref()).unwrap(), "c1".to_string()),
+                (col("c2", schema.as_ref()).unwrap(), "c2".to_string()),
+                (col("c3", schema.as_ref()).unwrap(), "c3".to_string()),
+            ],
+            input,
+        )?))
+    }
+
+    fn filter_exec(
+        schema: SchemaRef,
+        input: Arc<dyn ExecutionPlan>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(FilterExec::try_new(
+            Arc::new(BinaryExpr::new(
+                col("c3", schema.as_ref()).unwrap(),
+                Operator::Gt,
+                lit(0),
+            )),
+            input,
+        )?))
+    }
+
+    fn coalesce_batches_exec(input: Arc<dyn ExecutionPlan>) -> Arc<dyn ExecutionPlan> {
+        Arc::new(CoalesceBatchesExec::new(input, 8192))
+    }
+
+    fn coalesce_partitions_exec(
+        local_limit: Arc<dyn ExecutionPlan>,
+    ) -> Arc<dyn ExecutionPlan> {
+        Arc::new(CoalescePartitionsExec::new(local_limit))
+    }
+
+    fn repartition_exec(
+        streaming_table: Arc<dyn ExecutionPlan>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(RepartitionExec::try_new(
+            streaming_table,
+            Partitioning::RoundRobinBatch(8),
+        )?))
+    }
+}
```

Comment:
Can you create a scenario where LocalLimit's exist upper side of the plan (close to the root) and GlobalLimit's exist just after the source?